transformers_cfg/parser.py

```diff
@@ -10,22 +13,163 @@
 END_OF_GRAMMAR_MARKER = 0xFFFF
 TO_BE_FILLED_MARKER = 0
 REF_RULE_MARKER = 1
-LITERAL_MARKER = 2
 
 
 ########################
 # EBNF Grammar Parsing #
 ########################
 
 
+class Codable(ABC):
+    def serialize(self) -> List[int]:
+        raise NotImplementedError()
+
+    @classmethod
+    def deserialize(cls, data: List[int]) -> "Codable":
+        raise NotImplementedError()
+
+
+@dataclass
+class GrammarElement(Codable):
+    def is_terminated(self) -> bool:
+        raise NotImplementedError()
+    
+
+@dataclass
+class TerminatedElement(GrammarElement):
+    ranges: List[Tuple[int, int]]
+
+    def is_terminated(self) -> bool:
+        return True
+    
+    def serialize(self) -> List[int]:
+        outbuf = [len(self.ranges) * 2]
+        for range in self.ranges:
+            outbuf.extend(range)
+        return outbuf
+    
+    @classmethod
+    def from_range(cls, start: int, end: int) -> "GrammarElement":
+        return cls([(start, end)])
+
+
+@dataclass
+class ReferenceElement(GrammarElement):
+    reference_id: int
+
+    def is_terminated(self) -> bool:
+        return False
+    
+    def serialize(self) -> List[int]:
+        return [REF_RULE_MARKER, self.reference_id]
+    
+
+@dataclass
+class AlternativeElements(Codable):
+    buffer_elements: List[GrammarElement] = field(default_factory=list)
+    symbols: List[List[GrammarElement]] = field(default_factory=list)
+
+    def add_element(self, element: GrammarElement) -> None:
+        self.buffer_elements.append(element)
+
+    def add_symbol(self, symbol: List[GrammarElement]) -> None:
+        self.symbols.append(symbol)
+
+    def close_current_symbol(self) -> None:
+        self.symbols.append(self.buffer_elements)
+        self.buffer_elements = []
+
+    def serialize(self) -> List[int]:
+        if self.buffer_elements: self.close_current_symbol()
+        outbuf = [TO_BE_FILLED_MARKER]
+        for symbol in self.symbols:
+            for element in symbol:
+                outbuf.extend(element.serialize())
+        outbuf[0] = len(outbuf)
+        outbuf.append(END_OF_ALTERNATE_MARKER)
+        return outbuf
+
+
+@dataclass
+class GrammarRule(Codable):
+    id: int
+    name: str
+    alternatives: List[AlternativeElements] = field(default_factory=list)
+
+    def add_alternative(self, alternative: AlternativeElements) -> None:
+        self.alternatives.append(alternative)
+
+    def add_empty_alternative(self) -> AlternativeElements:
+        new_alternative = AlternativeElements()
+        self.alternatives.append(new_alternative)
+        return new_alternative
+
+    def serialize(self) -> List[int]:
+        outbuf = [self.id]
+        for alternative in self.alternatives:
+            outbuf.extend(alternative.serialize())
+        outbuf.append(END_OF_RULE_MARKER)
+        return outbuf
+
+
 class ParseState:
```

Comment:
I start to think that ParseState was really a bad name... How do you think `CompiledGrammar`?

Reply:
Let's change to `CompiledGrammar`

Reply:
After thinking about it, I think the `ParseState` is a good name because it represents the current parse state when parsing the grammar. But at the end, we can indeed use `CompiledGrammar`.